refactor: Extract RipeMd160

Empact opened the pull request:
To directly return a CRIPEMD160 hash from data.

Simplifies the call sites.

Empact committed:
refactor: Extract RIPEMD160

To directly return a CRIPEMD160 hash from data.

Incidentally, decoding this acronym:
* RIPEMD -> RIPE Message Digest
* RIPE -> RACE Integrity Primitives Evaluation
* RACE -> Research and Development in Advanced Communications Technologies in Europe

jimpo commented on src/hash.h:
```diff
@@ -89,7 +89,21 @@ inline uint256 Hash(const T1 p1begin, const T1 p1end,
+}
+
+/** Compute the 160-bit RIPEMD-160 hash of a vector. */
+inline uint160 RipeMd160(const std::vector<unsigned char>& vch)
```
Could make this a template function over things with `.data()` and `.size()` so that it works for `Span`s too. In which case you could also remove the other signature for `RipeMd160(ptr, size)` and just call `RipeMd160(Span<unsigned char>(ptr, size))`.

Empact replied:
I have a follow-up here: https://github.com/Empact/bitcoin/tree/base-blob-data that makes `RipeMd160` just a template and makes `base_blob` and descendants compatible by adding `base_blob::data()`. I figure it's best to put that up separately as it's mostly to do with `base_blob`.

jimpo reviewed (commented):
Concept ACK

Sjors commented:
Concept ACK. I noticed you're not touching the `CRIPEMD160` code in `script/interpreter.cpp`. I assume that's intentional?

This should have wallet tag I think, since is touch IsMine and signing code.

Why `hash.h`  rather than `ripemd160.h`? I'm a bit confused by the division of labor between hash.h and the various specific hash function files anyway.

Empact commented:
I have a follow-up here: 01a49bc61ef82517739909534804ab96192773e1 that makes `RipeMd160` just a template and makes `base_blob` and descendants compatible by adding `base_blob::data()`.

Empact commented:
@Sjors addressed your comments - now in `ripemd160.h`. Didn't touch `script/interpreter.cpp` both because the call pattern didn't fit, and to avoid touching consensus code.

Empact commented:
Rebased for #15638

Empact commented:
Rebased

theStack reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/15294/commits/b29e95c02b5a75681ee3b59d7dd8cceb85d8b27b

kristapsk reviewed (approved):
ACK b29e95c02b5a75681ee3b59d7dd8cceb85d8b27b

Sjors reviewed (approved):
Code review ACK b29e95c02b5a75681ee3b59d7dd8cceb85d8b27b

sipa commented on src/script/standard.h:
```diff
@@ -24,6 +24,7 @@ class CScriptID : public uint160
     CScriptID() : uint160() {}
     explicit CScriptID(const CScript& in);
     CScriptID(const uint160& in) : uint160(in) {}
+    CScriptID(const uint160&& in) : uint160(std::move(in)) {}
```
It doesn't do anything here, as it's equivalent to the const lvalue copy constructor.

Empact replied:
Thanks, removed this ineffective attempt at optimization.

Sjors commented:
re-ACK cc91980afd05796230e8ce95c622729accee1785

theStack commented:
re-ACK https://github.com/bitcoin/bitcoin/commit/cc91980afd05796230e8ce95c622729accee1785
(checked that only the const rvalue constructor has been removed since my previous ACK)

kristapsk reviewed (approved):
re-ACK ce38f54e66e8584cdd01b7d0dac76bbf010f0e2b

Empact commented:
Rebased

Empact commented:
Just noting this has had 4 ACKs.

kristapsk reviewed (approved):
re-ACK d7432d398846a4fb6a1ebf0a4d45b195c11acada

theStack reviewed (approved):
re-ACK d7432d398846a4fb6a1ebf0a4d45b195c11acada

kiminuo reviewed (approved):
crACK d7432d398846a4fb6a1ebf0a4d45b195c11acada

Empact commented on src/crypto/ripemd160.h:
```diff
@@ -25,4 +27,19 @@ class CRIPEMD160
 };
 
+/** Compute the 160-bit RIPEMD-160 hash of an array. */
+inline uint160 RipeMd160(const unsigned char* data, size_t size)
```
TIL: RIPEMD -> Research and Development in Advanced Communications Technologies in Europe Message Digest

I could go either way on this one - while all-caps is semantically correct for initialisms, capitalization has distinct meaning in code.

laanwj replied:
Maybe make this take a `Span<>` instead. I thint that might even make the second variant below unnecessary, as many containers automatically coerce to a span.

kristapsk reviewed (approved):
re-ACK cd7fe489df0c84d21c80b9df473e883e8a267f0a

theStack reviewed (approved):
re-ACK cd7fe489df0c84d21c80b9df473e883e8a267f0a

Checked via git range-diff that changes since my last ACK are only rebase-related.

Happy to also re-review if you decide to follow laanwj's suggestion of changing the `RipeMd160` interface to take a `Span<>` instead (https://github.com/bitcoin/bitcoin/pull/15294#discussion_r705283491).

MarcoFalke commented on src/crypto/ripemd160.h:
```diff
@@ -28,18 +29,11 @@ class CRIPEMD160
 
-/** Compute the 160-bit RIPEMD-160 hash of an object. */
-template <typename T>
-inline uint160 RipeMd160(const T& container)
```
Instead of adding code and removing it in the next commit, it might be better to squash the commits. This will reduce review burden for reviewers starting from scratch as well as reduce the git history, blame depth and clone size.

jonatack commented on src/wallet/rpcdump.cpp:
```diff
@@ -910,8 +911,7 @@ static std::string RecurseImportData(const CScript& script, ImportData& import_d
         uint256 fullid(solverdata[0]);
-        CScriptID id;
-        CRIPEMD160().Write(fullid.begin(), fullid.size()).Finalize(id.begin());
+        CScriptID id{RipeMd160(fullid)};
```
suggestion `s/fullid/full_id/`, or better yet remove it
```diff
-        uint256 fullid(solverdata[0]);
-        CScriptID id{RipeMd160(fullid)};
+        CScriptID id{RipeMd160(solverdata[0])};
```

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3816,10 +3815,8 @@ class DescribeWalletAddressVisitor
-        hasher.Write(id.begin(), 32).Finalize(hash.begin());
-        if (provider && provider->GetCScript(CScriptID(hash), subscript)) {
+        CScriptID scriptid{RipeMd160(id)};
+        if (provider && provider->GetCScript(scriptid, subscript)) {
```
suggestion `s/scriptid/script_id/`
```diff
@@ -3814,9 +3814,9 @@ public:
     UniValue operator()(const WitnessV0ScriptHash& id) const
     {
         UniValue obj(UniValue::VOBJ);
+        CScriptID script_id{RipeMd160(id)};
         CScript subscript;
-        CScriptID scriptid{RipeMd160(id)};
-        if (provider && provider->GetCScript(scriptid, subscript)) {
+        if (provider && provider->GetCScript(script_id, subscript)) {
             ProcessSubScript(subscript, obj);
         }
         return obj;
```

jonatack commented on src/crypto/ripemd160.h:
```diff
@@ -8,6 +8,9 @@
 #include <stdlib.h>
 
+#include <uint256.h> // For uint160
+#include <span.h> // For Span
```
suggestion, sort
```diff
+#include <span.h>
 #include <uint256.h> // For uint160
-#include <span.h> // For Span
```

jonatack reviewed (commented):
Nice refactoring/cleanup.

Code review ACK cd7fe489df0c84d21c80b9df473e883e8a267f0a and debug build is clean

Some minor suggestions for your consideration if you retouch.

jonatack commented:
Perhaps `s/moveonly/refactoring/` in the PR title.

stratospher reviewed (commented):
ACK ca55641. This PR defines RIPEMD160() as an inline function in src/hash.h and makes the function calls more intuitive.

achow101 commented:
Are you still working on this?

Empact commented:
@achow101 Rebased to address conflict with #26158

achow101 commented:
Two more places in `src/wallet/rpc/coins.cpp` could be updated:

```diff
diff --git a/src/wallet/rpc/coins.cpp b/src/wallet/rpc/coins.cpp
index 82642194c20..250b580c37c 100644
--- a/src/wallet/rpc/coins.cpp
+++ b/src/wallet/rpc/coins.cpp
@@ -679,8 +679,7 @@ RPCHelpMan listunspent()
                             CHECK_NONFATAL(extracted);
                             // Also return the witness script
                             const WitnessV0ScriptHash& whash = std::get<WitnessV0ScriptHash>(witness_destination);
-                            CScriptID id;
-                            CRIPEMD160().Write(whash.begin(), whash.size()).Finalize(id.begin());
+                            CScriptID id{RIPEMD160(whash)};
                             CScript witnessScript;
                             if (provider->GetCScript(id, witnessScript)) {
                                 entry.pushKV("witnessScript", HexStr(witnessScript));
@@ -689,8 +688,7 @@ RPCHelpMan listunspent()
                     }
                 } else if (scriptPubKey.IsPayToWitnessScriptHash()) {
                     const WitnessV0ScriptHash& whash = std::get<WitnessV0ScriptHash>(address);
-                    CScriptID id;
-                    CRIPEMD160().Write(whash.begin(), whash.size()).Finalize(id.begin());
+                    CScriptID id{RIPEMD160(whash)};
                     CScript witnessScript;
                     if (provider->GetCScript(id, witnessScript)) {
                         entry.pushKV("witnessScript", HexStr(witnessScript));
```

achow101 commented:
ACK 6879be691bf636a53208ef058f2ebe18bfa8017c

theStack reviewed (approved):
re-ACK 6879be691bf636a53208ef058f2ebe18bfa8017c

maflcko reviewed (approved):
review ACK 6879be691bf636a53208ef058f2ebe18bfa8017c  🏔

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 6879be691bf636a53208ef058f2ebe18bfa8017c  🏔
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiT1wv+JeOoQOIiy2NUd4vV1htnx+kXGz/gFFoDnqCY0M9sj6LdXfmwLf3EYpQA
fKhnS86I49ms/985+Z9jviAiwZ6iA9ViVZjmoLhcfPqUi8l/cDInfml8UVwLI2dj
0aXCAqXRex56hieA3q8zHCQ3L12UPyD/pJvdVGIDim18llMB5lxld7gT7eC3fhyO
i+jo6x86YDdxo7+OgwMXjSb/59IPK4bR/B10lBS64AGLKD1tSCpU2X/hQkUnNsut
Q/RCJipWDBL5cepJwCR/h1XO4C2KFzYEw998cYm0ZnLLivKTKimrHeDgQfJ3akfZ
m3YRylDN8OeJ/Fnwr7JuPisTBvisVm7wfFIF30RW8AgPF/5WFR2iddF7Awlf83Qr
AMnEcCnM6mMCO1xxrxR92NjnEqXsqlTqf6f0ygNuCiFIA5xyADUvT3B+dMiT0eg5
0M6xp/ppnrP5AA8Nc9m+FZ9tEDOwiCh/9cYeZAZyyruEhw1JgjZyUAgU7qMD2cr4
Ev+BKUFf
=XLKt
-----END PGP SIGNATURE-----
```


</details>
